editor/src/components/canvas/remix/remix-utils.tsx

```diff
@@ -99,3 +128,397 @@ export function getRoutesFromRouteManifest(
     return []
   }
 }
+
+export function createAssetsManifest(routes: RouteManifest<EntryRoute>): AssetsManifest {
+  return {
+    entry: { imports: [], module: '' },
+    url: '/',
+    version: '1',
+    routes: routes,
+  }
+}
+
+export interface RouteModuleCreator {
+  filePath: string
+  executionScopeCreator: (projectContents: ProjectContentTreeRoot) => ExecutionScope
+}
+
+export interface RouteIdsToModuleCreators {
+  [routeId: string]: RouteModuleCreator
+}
+
+export interface RouteModulesWithRelativePaths {
+  [routeId: string]: {
+    relativePath: ElementPath
+    filePath: string
+  }
+}
+
+export interface GetRoutesAndModulesFromManifestResult {
+  routeModuleCreators: RouteIdsToModuleCreators
+  routes: Array<DataRouteObject>
+  routeModulesToRelativePaths: RouteModulesWithRelativePaths
+  routingTable: RemixStaticRoutingTable
+}
+
+export function getDefaultExportNameAndUidFromFile(
+  projectContents: ProjectContentTreeRoot,
+  filePath: string,
+): { name: string; uid: string } | null {
+  const file = getProjectFileByFilePath(projectContents, filePath)
+  if (file == null || file.type != 'TEXT_FILE' || file.lastParseSuccess == null) {
+    return null
+  }
+
+  const defaultExportName =
+    file.lastParseSuccess.exportsDetail.find(
+      (e): e is ExportDefaultFunctionOrClass => e.type === 'EXPORT_DEFAULT_FUNCTION_OR_CLASS',
+    )?.name ?? null
+
+  if (defaultExportName == null) {
+    return null
+  }
+
+  const elementUid = file.lastParseSuccess.topLevelElements.find(
+    (t): t is UtopiaJSXComponent =>
+      t.type === 'UTOPIA_JSX_COMPONENT' && t.name === defaultExportName,
+  )?.rootElement.uid
+  if (elementUid == null) {
+    return null
+  }
+
+  return { name: defaultExportName, uid: elementUid }
+}
+
+export function getDefaultExportedTopLevelElement(file: TextFile): JSXElementChild | null {
+  if (file.fileContents.parsed.type !== 'PARSE_SUCCESS') {
+    return null
+  }
+
+  const defaultExportName =
+    file.fileContents.parsed.exportsDetail.find(
+      (e): e is ExportDefaultFunctionOrClass => e.type === 'EXPORT_DEFAULT_FUNCTION_OR_CLASS',
+    )?.name ?? null
+
+  if (defaultExportName == null) {
+    return null
+  }
+
+  return (
+    file.fileContents.parsed.topLevelElements.find(
+      (t): t is UtopiaJSXComponent =>
+        t.type === 'UTOPIA_JSX_COMPONENT' && t.name === defaultExportName,
+    )?.rootElement ?? null
+  )
+}
+
+function addLoaderAndActionToRoute(
+  routes: DataRouteObject[],
+  routeId: string,
+  loader: LoaderFunction | undefined,
+  action: ActionFunction | undefined,
+) {
+  routes.forEach((route) => {
+    if (route.id === routeId) {
+      route.action = action
+      route.loader = loader
+    } else {
+      addLoaderAndActionToRoute(route.children ?? [], routeId, loader, action)
+    }
+  })
+}
+
+function findOutletAmongJSXElementChildren(
+  parentUid: string,
+  isOutlet: (e: JSXElementChild) => boolean,
```

Comment:
Seems like this is for finding some subset of all the outlets?
```suggestion
  outletPredicate: (e: JSXElementChild) => boolean,
```

Reply:
I think this is really just used to find outlets, so I'm not sure why is this function injected at all. @bkrmendy ?

Reply:
So I believe the function is just passed in so we don't have to pass in all the information necessary to find outlets. But then basically this is a general `findAmongJSXElementChildren`, and the `isOutlet` can be any kind of predicate.

Reply:
Fixed in caddafb09445da6df2a48879e0cc405de3be0994